File: 14 robots - book ofi/python/book_ofi.py
import os
import sys

sys.path.insert(0, os.path.join(os.path.dirname(__file__), "..", "..", "helpers", "python"))
import tse_helpers as H

tse = H.tse

_message_counter = {"value": 0}


def new_order(ts_nanoseconds, side, price, quantity):
    _message_counter["value"] += 1
    ident = _message_counter["value"]
    return tse.make_book_message(tse.BookMessageKind.New, ts_nanoseconds, ident, ident, price, quantity, side)


def trade_tick(ts_nanoseconds, price):
    return tse.TseTickTrade(ts_nanoseconds, price, 1.0, int(tse.Side.Trade))


def market_params(txn_side, pos_side):
    return tse.make_rule_params(
        tse.Quantity.Fixed, 100.0,
        tse.Price.Market, 0.0, 0.0, 0.0,
        txn_side, pos_side, tse.Tif.Day, 10,
    )


# A laboratory feed rather than a recording: every step adds exactly the size that drags the
# book onto the next target imbalance, so the sequence of signals is known before the run.
def make_imbalance_feed(targets, prices, bid_price, ask_price, initial_bid_quantity):
    base = 1000000000
    half = 500000000
    book_feed = []
    price_feed = []
    sb = initial_bid_quantity
    sa = 0.0
    step = 0

    price_feed.append(trade_tick(base - half, prices[0]))
    book_feed.append(new_order(base * (2 * step + 1), tse.Side.Long, bid_price, initial_bid_quantity))
    price_feed.append(trade_tick(base * (2 * step + 1) + half, prices[0]))
    step += 1
    for i, t in enumerate(targets):
        curr = (sb - sa) / (sb + sa)
        add_bid = t > curr
        if add_bid:
            delta = (sa * (1.0 + t) / (1.0 - t)) - sb
            sb += delta
            side = tse.Side.Long
            level_price = bid_price
        else:
            delta = (sb * (1.0 - t) / (1.0 + t)) - sa
            sa += delta
            side = tse.Side.Short
            level_price = ask_price
        book_feed.append(new_order(base * (2 * step + 1), side, level_price, delta))
        price_feed.append(trade_tick(base * (2 * step + 1) + half, prices[i + 1]))
        step += 1
    return book_feed, price_feed


def run_scenario(level, kind, book_feed, price_feed):
    symbol = "BOOK" + level
    long_label = level + "Long"
    short_label = level + "Short"

    account = H.account("Ofi" + level, tse.StorageRegime.Mem, tse.LogLevel.Off)
    book_mkt = account.create_market("book", tse.MdType.Book)
    price_mkt = account.create_market("price", tse.MdType.Trade)
    account.create_simulator("sim", H.simulator_options(), 64, -1)
    account.add_contract(symbol, 1, tse.Instrument.Equity, tse.Underlying.Equity, tse.Venue.Undefined, 50000)
    book = account.create_book(symbol, kind)

    # Resting-size imbalance over the whole book, with a fifth of extra size on one side
    # taken as pressure worth joining; the opposite reading closes the position. The 4 is the
    # length of the input's own storage, not a book depth.
    account.add_input_book_imbalance(level, 4, tse.Duration.Nanoseconds, book, book_mkt, [symbol])
    account.add_pattern_threshold(long_label, tse.Duration.Nanoseconds, [level], tse.Cmp.Ge, 0.2)
    account.add_pattern_threshold(short_label, tse.Duration.Nanoseconds, [level], tse.Cmp.Le, -0.2)

    account.add_rule_market("EnterLong", tse.RuleType.Entry, market_params(tse.Side.Long, tse.Side.Neutral), long_label, symbol)
    account.add_rule_market("ExitLong", tse.RuleType.Exit, market_params(tse.Side.Short, tse.Side.Long), short_label, symbol)
    account.add_rule_market("EnterShort", tse.RuleType.Entry, market_params(tse.Side.Short, tse.Side.Neutral), short_label, symbol)
    account.add_rule_market("ExitShort", tse.RuleType.Exit, market_params(tse.Side.Long, tse.Side.Short), long_label, symbol)
    account.add_robot("Robot", ["EnterLong", "ExitLong", "EnterShort", "ExitShort"])
    account.portfolio_subscribe(price_mkt, symbol)
    account.start("Robot")

    price_mkt.push_trade_by_name(symbol, price_feed[0])
    for i, message in enumerate(book_feed):
        book_mkt.push_book_by_name(symbol, message)
        price_mkt.push_trade_by_name(symbol, price_feed[i + 1])

    summary = account.get_summary()
    account.close()
    return summary


def main():
    targets = [-0.4, -0.5, 0.4, 0.5, -0.4, -0.5, 0.4, 0.5, -0.4, -0.5, 0.4]
    prices = [100, 102, 102, 100, 100, 102, 102, 100, 100, 102, 102, 100]
    book_feed, price_feed = make_imbalance_feed(targets, prices, 99.95, 100.05, 12.0)

    levels = ["L1", "L2", "L3"]
    kinds = [tse.BookLevelKind.L1, tse.BookLevelKind.L2, tse.BookLevelKind.L3]

    # The same strategy over the same feed, on the three book depths a venue may publish. The
    # feed touches one price per side, so the depth changes what is stored, never the verdict.
    results = []
    for i in range(len(levels)):
        results.append(run_scenario(levels[i], kinds[i], book_feed, price_feed))

    print("netProfit L1={:.4f} L2={:.4f} L3={:.4f}".format(
        results[0].totalNetProfit, results[1].totalNetProfit, results[2].totalNetProfit))
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: 14 robots - book ofi/cpp/book_ofi.cpp
#include "tse_helpers.hpp"

#include <cstdint>
#include <cstdio>
#include <string>
#include <vector>

namespace {

	struct BookFeed final {
		std::vector<tse::BookMessage> book;
		std::vector<tse::TradeTick> price;
	};

	std::uint64_t messageCounter {0};

	tse::BookMessage newOrder
	(
		std::int64_t tsNanoseconds,
		tse::Side side,
		double price,
		double quantity
	)
	{
		tse::BookMessage message;
		message.kind = tse::BookMessageKind::new_;
		message.tsNanoseconds = tsNanoseconds;
		message.messageId = ++messageCounter;
		message.orderId = message.messageId;
		message.price = price;
		message.quantity = quantity;
		message.txnSide = side;
		return message;
	}

	tse::TradeTick tradeTick(std::int64_t tsNanoseconds, double price)
	{
		return tse::TradeTick {tsNanoseconds, price, 1.0, tse::Side::trade};
	}

	tse::RuleParams marketParams
	(
		tse::Side txnSide,
		tse::Side posSide
	)
	{
		return tse::RuleParams
		{
			tse::QuantityMode::fixed, 100.0,
			tse::PriceType::market, 0.0, 0.0, 0.0,
			txnSide, posSide, tse::Tif::day, 10
		};
	}

	// A laboratory feed rather than a recording: every step adds exactly the size that drags the
	// book onto the next target imbalance, so the sequence of signals is known before the run.
	BookFeed makeImbalanceFeed
	(
		std::vector<double> const& targets,
		std::vector<double> const& prices,
		double bidPrice,
		double askPrice,
		double initialBidQuantity
	)
	{
		std::int64_t constexpr
		base {1000000000LL},
		half {500000000LL};
		BookFeed feed;
		double
		sb {initialBidQuantity},
		sa {0.0};
		std::int64_t step {0};

		feed.price.push_back(tradeTick(base - half, prices[0]));
		feed.book.push_back(newOrder(base * (2 * step + 1), tse::Side::long_, bidPrice, initialBidQuantity));
		feed.price.push_back(tradeTick(base * (2 * step + 1) + half, prices[0]));
		++step;
		for (std::size_t i {0}; i < targets.size(); ++i) {
			double const
			t {targets[i]},
			curr {(sb - sa) / (sb + sa)};
			bool const addBid {t > curr};
			double const delta
			{
				addBid
					? (sa * (1.0 + t) / (1.0 - t)) - sb
					: (sb * (1.0 - t) / (1.0 + t)) - sa
			};
			if (addBid) {
				sb += delta;
			}
			else {
				sa += delta;
			}
			feed.book.push_back(newOrder(base * (2 * step + 1), addBid ? tse::Side::long_ : tse::Side::short_, addBid ? bidPrice : askPrice, delta));
			feed.price.push_back(tradeTick(base * (2 * step + 1) + half, prices[i + 1]));
			++step;
		}
		return feed;
	}

	tse::Summary runScenario
	(
		std::string const& level,
		tse::BookLevelKind kind,
		BookFeed const& feed
	)
	{
		std::string const
		symbol {"BOOK" + level},
		longLabel {level + "Long"},
		shortLabel {level + "Short"};

		tse::Account account {"Ofi" + level, tse::StorageRegime::mem};
		tse::Market const bookMkt {account.createMarket("book", tse::MdType::book)};
		tse::Market const priceMkt {account.createMarket("price", tse::MdType::trade)};
		account.createSimulator("sim", helpers::simulatorConfig(), 64, -1);
		account.addContract(symbol, 1, tse::Instrument::equity, tse::Underlying::equity, tse::Venue::undefined, 50000);
		tse::Book const book {account.createBook(symbol, kind)};

		// Resting-size imbalance over the whole book, with a fifth of extra size on one side
		// taken as pressure worth joining; the opposite reading closes the position. The 4 is the
		// length of the input's own storage, not a book depth.
		account.addInputBookImbalance(level, 4, tse::Duration::nanoseconds, book, bookMkt, {symbol});
		account.addPatternThreshold(longLabel, tse::Duration::nanoseconds, {level}, tse::Cmp::ge, 0.2);
		account.addPatternThreshold(shortLabel, tse::Duration::nanoseconds, {level}, tse::Cmp::le, -0.2);

		account.addRuleMarket("EnterLong", tse::RuleType::entry, marketParams(tse::Side::long_, tse::Side::neutral), longLabel, symbol);
		account.addRuleMarket("ExitLong", tse::RuleType::exit, marketParams(tse::Side::short_, tse::Side::long_), shortLabel, symbol);
		account.addRuleMarket("EnterShort", tse::RuleType::entry, marketParams(tse::Side::short_, tse::Side::neutral), shortLabel, symbol);
		account.addRuleMarket("ExitShort", tse::RuleType::exit, marketParams(tse::Side::long_, tse::Side::short_), longLabel, symbol);
		account.addRobot("Robot", {"EnterLong", "ExitLong", "EnterShort", "ExitShort"});
		account.portfolioSubscribe(priceMkt, symbol);
		account.start("Robot");

		priceMkt.pushTrade(symbol, feed.price[0]);
		for (std::size_t i {0}; i < feed.book.size(); ++i) {
			bookMkt.pushBook(symbol, feed.book[i]);
			priceMkt.pushTrade(symbol, feed.price[i + 1]);
		}

		return account.getSummary();
	}

}

int main()
{
	tse::setLogLevel(tse::LogLevel::none);

	std::vector<double> const targets
	{
		-0.4, -0.5, 0.4, 0.5, -0.4, -0.5, 0.4, 0.5, -0.4, -0.5, 0.4
	};
	std::vector<double> const prices
	{
		100, 102, 102, 100, 100, 102, 102, 100, 100, 102, 102, 100
	};
	BookFeed const feed {makeImbalanceFeed(targets, prices, 99.95, 100.05, 12.0)};

	std::vector<std::string> const levels {"L1", "L2", "L3"};
	std::vector<tse::BookLevelKind> const kinds
	{
		tse::BookLevelKind::l1, tse::BookLevelKind::l2, tse::BookLevelKind::l3
	};

	// The same strategy over the same feed, on the three book depths a venue may publish. The
	// feed touches one price per side, so the depth changes what is stored, never the verdict.
	std::vector<tse::Summary> results;
	for (std::size_t i {0}; i < levels.size(); ++i) {
		results.push_back(runScenario(levels[i], kinds[i], feed));
	}

	std::printf
	(
		"netProfit L1=%.4f L2=%.4f L3=%.4f\n",
		results[0].totalNetProfit,
		results[1].totalNetProfit,
		results[2].totalNetProfit
	);
	return 0;
}
